File: tests/cpp/src/TestWaitStorage.cpp
/*
 * CAMEO
 *
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under BSD 3-Clause and GPL-v3 as described in license files.
 * You may not use this work except in compliance with the Licences.
 *
 */

#include <iostream>
#include <string>
#include <vector>
#include <sstream>
#include <cameo/api/cameo.h>

using namespace std;
using namespace cameo;

int main(int argc, char *argv[]) {

	int numberOfTimes = 1;

	if (argc > 2) {
		numberOfTimes = stoi(argv[1]);
	}

	This::init(argc, argv);

	int options = 0;
	bool useProxy = false;
	string endpoint = "tcp://localhost:11000";
	if (argc > 3) {
		useProxy = (string(argv[2]) == "true");
	}
	if (useProxy) {
		options |= option::USE_PROXY;
		endpoint = "tcp://localhost:12000";
	}

	unique_ptr<Server> server = Server::create(endpoint, options);
	server->init();

	// Define the shared key.
	string key = "eu.ill.cameo.test.testkey";

	// loop the number of times.
	for (int i = 0; i < numberOfTimes; ++i) {

		// start the application.
		unique_ptr<App> app = server->start("waitstoragecpp");

		// Define a KeyValue.
		KeyValue keyValue(key);

		// waitFor blocks.
		app->waitFor(keyValue);
		cout << "storage event " << keyValue.getStatus() << " " << keyValue.getValue() << endl;

		// Get the key value.
		cout << "key value " << app->getCom().getKeyValue(key) << endl;

		// waitFor blocks.
		app->waitFor(keyValue);
		cout << "storage event " << keyValue.getStatus() << " " << keyValue.getValue() << endl;

		state::Value state = app->waitFor();

		cout << "finished the application " << *app << " with state " << toString(state) << endl;
	}

	This::terminate();

	return 0;
}

File: tests/java/src/eu/ill/cameo/test/TestWaitStorage.java
/*
 * CAMEO
 *
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under BSD 3-Clause and GPL-v3 as described in license files.
 * You may not use this work except in compliance with the Licences.
 *
 */

package eu.ill.cameo.test;

import java.util.Date;

import eu.ill.cameo.api.base.App;
import eu.ill.cameo.api.base.KeyValue;
import eu.ill.cameo.api.base.Option;
import eu.ill.cameo.api.base.Server;
import eu.ill.cameo.api.base.This;


public class TestWaitStorage {

	public static void main(String[] args) {
		
		This.init(args);
		
		int numberOfTimes = 1;
		
		if (args.length > 1) {
			numberOfTimes = Integer.parseInt(args[0]);
		}
		
		boolean useProxy = false;
		String endpoint = "tcp://localhost:11000";
		if (args.length > 2) {
			useProxy = Boolean.parseBoolean(args[1]);
		}
		if (useProxy) {
			endpoint = "tcp://localhost:12000";
		}
		
		Server server = Server.create(endpoint, (useProxy ? Option.USE_PROXY : 0));
		server.init();
		
		Date d = new Date();
		
		String key = "eu.ill.cameo.test.testkey";
		
		try {
			// Loop the number of times.
			for (int i = 0; i < numberOfTimes; ++i) {
			
				// Start the application.
				App application = server.start("waitstoragejava");
	
				// Define a KeyValue.
				KeyValue keyValue = new KeyValue(key);
				
				application.waitFor(keyValue);
				System.out.println("Storage event " + keyValue.getStatus() + " " + keyValue.getValue());
				
				// Get the key value.
				try {
					System.out.println("Key value " + application.getCom().getKeyValue(key));
				}
				catch (Exception e) {
				}
				
				application.waitFor(keyValue);
				System.out.println("Storage event " + keyValue.getStatus() + " " + keyValue.getValue());
								
				application.waitFor();
				System.out.println("Finished application " + application);
			}
		}
		finally {
			server.terminate();
			This.terminate();
		}
		
		System.out.println("Finished the application after " + ((new Date()).getTime() - d.getTime()) + "ms");
	}

}